p2p/server.py

```diff
@@ -233,24 +237,37 @@ def _make_syncer(self, peer_pool: PeerPool) -> BaseService:
         return FullNodeSyncer(
             self.chain, self.chaindb, self.base_db, peer_pool, self.cancel_token)
 
-    def _make_peer_pool(self, discovery: DiscoveryProtocol) -> PeerPool:
+    def _make_peer_pool(self, fut_discovery: 'asyncio.Future[DiscoveryProtocol]') -> PeerPool:
         # This method exists only so that ShardSyncer can provide a different implementation.
         return self.peer_pool_class(
             self.peer_class,
             self.headerdb,
             self.network_id,
             self.privkey,
-            discovery,
+            fut_discovery,
             max_peers=self.max_peers,
         )
 
-    async def _run(self) -> None:
-        self.logger.info("Running server...")
-        upnp_dev = await self._discover_upnp_device()
+    async def _make_discovery(self,
+                              upnp_device: Awaitable[upnpclient.Device]) -> DiscoveryProtocol:
+
+        upnp_dev = await upnp_device
         external_ip = '0.0.0.0'
         if upnp_dev is not None:
-            external_ip = upnp_dev.WANIPConn1.GetExternalIPAddress()['NewExternalIPAddress']
             await self._add_nat_portmap(upnp_dev)
+            external_ip = upnp_dev.WANIPConn1.GetExternalIPAddress()['NewExternalIPAddress']
+
+        addr = Address(external_ip, self.port, self.port)
```

Comment:
Just realized this will collide with our `eth_typing.Address`.  Just pointing it out.

Reply:
`Address` is a pretty generic name so I expect it to exist in many different modules. That said, in this module only `kademlia.Address` is used. Do you want me to change that to `kademlia.Address` to make that more explicit?

Reply:
No, just wanted to point it out as we'll probably make the mistake eventually.